File: Mathematics/Inclusion-Exclusion Principle/atcoder/B31/B31.py
# 以下に、**Python（CPython 3.11.4）** を用いて問題を解く解法を示します。

# ---

# ## ✅ 問題概要の再確認

# * 与えられた整数 `N (1 ≤ N ≤ 10^12)` に対して、
# * `1 以上 N 以下` の整数のうち、`3, 5, 7` のいずれかで割り切れる数の個数を求める。

# ---

# ## ✅ 解法：包除原理（Inclusion-Exclusion）

# ```text
# |A ∪ B ∪ C| = |A| + |B| + |C| - |A ∩ B| - |A ∩ C| - |B ∩ C| + |A ∩ B ∩ C|
# ```

# ---

# ## ✅ Python コード（型付き + コメントあり）

# ```python
import sys

def count_multiples(N: int) -> int:
    """
    N以下の整数のうち、3, 5, 7 のいずれかで割り切れる整数の個数を返す

    Parameters:
    - N (int): 1以上10^12以下の整数

    Returns:
    - int: 条件を満たす整数の個数
    """

    def div(x: int) -> int:
        return N // x

    # Inclusion-Exclusion Principle
    count: int = (
        div(3) + div(5) + div(7)
        - div(3 * 5) - div(3 * 7) - div(5 * 7)
        + div(3 * 5 * 7)
    )
    return count

def main() -> None:
    """
    標準入力からNを受け取り、条件を満たす整数の個数を出力する
    """
    input_data: str = sys.stdin.read().strip()
    N: int = int(input_data)
    result: int = count_multiples(N)
    print(result)

if __name__ == "__main__":
    main()

# ## ✅ 実行例

# ```bash
# echo 10 | python3 main.py
# # 出力: 6
# ```

# ---

# ## ✅ 処理時間・メモリ消費

# | 項目        | 内容                                    |
# | --------- | ------------------------------------- |
# | **計算量**   | O(1)（整数除算を7回）                         |
# | **実行時間**  | 約 0.01ms ～ 0.1ms 程度（Nに依存しない）          |
# | **メモリ使用** | 数個の整数変数（< 1KB）                        |
# | **精度**    | Python の `int` は任意精度 → `10^12` 以上でも安全 |

# ---

# ## ✅ テスト例

# | 入力値 `N`        | 出力            | 備考           |
# | -------------- | ------------- | ------------ |
# | `10`           | `6`           | 3,5,6,7,9,10 |
# | `210`          | `114`         |              |
# | `100000000000` | `54285714286` | 任意精度でも高速     |

# ---

# ご希望があれば、**ナイーブなO(N)解法やベン図解説付きのPythonコード**も併せて提供可能です。

File: Mathematics/Inclusion-Exclusion Principle/atcoder/B31/B31.ts
// 以下は、**TypeScript 5.1 + Node.js 18.16.1** で `1 以上 N 以下の整数のうち、3, 5, 7 のいずれかで割り切れる数` の個数を求めるプログラムです。

// ---

// ### ✅ 解法：**包除原理（Inclusion-Exclusion Principle）**

// * N 以下の数で 3, 5, 7 のいずれかで割り切れる数を数えるには：

//   * `N / 3 + N / 5 + N / 7`
//   * `- N / 15 - N / 21 - N / 35`（二重カウント除外）
//   * `+ N / 105`（三重カウントを再加算）

// ---

// ### ✅ TypeScript コード（`fs` + `BigInt` 使用）

// ```ts
import * as fs from 'fs';

/**
 * N以下の整数のうち、3,5,7のいずれかで割り切れる整数の個数を返す
 * @param {bigint} N - 1以上10^12以下の整数
 * @returns {bigint} - 条件を満たす整数の個数
 */
function countMultiples(N: bigint): bigint {
  const div = (x: bigint): bigint => N / x;

  // Inclusion-Exclusion Principle
  const count =
    div(3n) + div(5n) + div(7n)
    - div(3n * 5n) - div(3n * 7n) - div(5n * 7n)
    + div(3n * 5n * 7n);

  return count;
}

/**
 * メイン処理（標準入力から読み込み、結果を出力）
 */
function main(): void {
  const input: string = fs.readFileSync('/dev/stdin', 'utf8').trim();
  const N: bigint = BigInt(input);

  const result: bigint = countMultiples(N);
  console.log(result.toString()); // BigInt型を文字列で出力
}

main();

// ### ✅ 処理時間・メモリ使用量

// | 要素      | 内容                     |
// | ------- | ---------------------- |
// | **計算量** | `O(1)`（定数回の除算のみ）       |
// | **メモリ** | `~100バイト`（BigInt変数数個）  |
// | **速度**  | 約 0.1ms 以下（I/O 除く）     |
// | **安全性** | `BigInt`使用により10^12まで対応 |

// ---

// ### ✅ 実行方法

// #### コマンド：

// ```bash
// tsc main.ts
// node main.js < input.txt
// ```

// または `ts-node` を使う場合：

// ```bash
// ts-node main.ts < input.txt
// ```

// ---

// ### ✅ 入力例と出力例の確認

// #### 入力：

// ```
// 210
// ```

// #### 出力：

// ```
// 114
// ```

// ---

// 必要であれば、ナイーブな `for` ループ実装（O(N)）や、追加のデバッグ出力付きのバージョンも提供できます。
